Refactor user and channel storage

jvperrin opened the pull request:
This simplifies the storage of the mappings of users and channels to store them in the `IRCBot` class instead of passing them around everywhere. As @kkuehler mentioned on IRC/Slack to me already, the channel mapping stuff could be done even earlier, in `main.py` or something, but the user mapping can't be done until the factories start, because it maps user IDs to bots, which aren't available until later.

kkuehler commented on slackbridge/bots.py:
```diff
@@ -62,8 +67,8 @@ def resolve(self):
             channel_id = self.raw_message['channel']
-            if channel_id in self.channels:
-                channel_name = self.channels[channel_id]['name']
+            if channel_id in IRCBot.channels.keys():
```
What is the benefit of doing this versus:
```python
if channel_id in IRCBot.channels
```
That is the convention we've been following for this project.

jvperrin replied:
That would work too and is cleaner, thanks for catching that.

kkuehler commented on slackbridge/bots.py:
```diff
@@ -229,17 +227,21 @@ def signedOn(self):
+            self.joined_channels.append(channel_name)
 
     def part_channel(self, channel_name):
         self.leave('#{}'.format(channel_name))
```
Should we check if the channel is in ```joined_channels``` before calling the ```leave``` method? Or could this lead to unexpected behavior if we, for example, joined and parted a channel in quick succession?

kkuehler replied:
Twisted also has a ```joined``` method that might be worth looking into. It's called when a user finishes joining a channel. That seems like a more logical place to append to ```joined_channels```.

jvperrin replied:
Given `leave` just calls `PART` under the hood, I don't think it really matters that much. I think if the bot isn't in the channel but tries to leave, it won't cause any problems, although I'd need to actually test that. The `joined_channel` stuff is really just an optimization right now anyway, so that all the bots don't join all the IRC channels. I do think your suggestion is fine, my main concern is that the `joined_channels` list gets out of date and then a bot can't leave (or worse, join) a channel any more.

kkuehler commented on slackbridge/factories.py:
```diff
@@ -71,24 +72,24 @@ def instantiate_bot(self, user):
         self.nickserv_password = nickserv_pw
 
-        for channel in channels:
+        for channel_id, channel in IRCBot.channels.items():
```
```python
for _, channel in IRCBot.channels.items()
```

kkuehler replied:
Or, better yet, ```IRCBot.channels.values()```. Doesn't seem necessary to have ```channel_id```.

jvperrin replied:
Yeah, using `.values()` would be better here, thanks.

kkuehler commented on slackbridge/bots.py:
```diff
@@ -229,17 +227,21 @@ def signedOn(self):
     def part_channel(self, channel_name):
         self.leave('#{}'.format(channel_name))
+        if channel_name in self.joined_channels:
+            self.joined_channels.remove(channel_name)
```
Same thing here. The ```left``` method seems like a better place to remove channels. Then you can safely do
```python
if channel_name in self.joined_channels:
    self.leave('#{}'.format(channel_name)
```
in ```part_channel```

jvperrin commented:
Ok, @kkuehler, I addressed your comments, and then also moved the `SlackMessage` class to its own module, since I agree with you that it doesn't need to be in the bot file.
